### PR TITLE
8343218: Disable allocating interface and abstract classes in non-class metaspace

### PR DESCRIPTION
**Pull request by coleenp**
Change the code to allocate all Klasses to class-metaspace until c2 dependencies are understood and fixed, and more urgent need arises when compressing klass pointers to smaller numbers of bits.
Tested with tier1-4.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8343218](https://bugs.openjdk.org/browse/JDK-8343218): Disable allocating interface and abstract classes in non-class metaspace (**Sub-task** - P4)

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) 🔄 Re-review required (review applies to [a30081ad](https://git.openjdk.org/jdk/pull/21769/files/a30081ade78a45d2ade632833982c065ad939456))

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21769/head:pull/21769` \
`$ git checkout pull/21769`

Update a local copy of the PR: \
`$ git checkout pull/21769` \
`$ git pull https://git.openjdk.org/jdk.git pull/21769/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21769`

View PR using the GUI difftool: \
`$ git pr show -t 21769`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21769.diff">https://git.openjdk.org/jdk/pull/21769.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21769#issuecomment-2444904309)

### CODE REVIEW

File: src/hotspot/share/oops/instanceKlass.cpp [hunk 1]
```diff
@@ -454,7 +454,10 @@ InstanceKlass* InstanceKlass::allocate_instance_klass(const ClassFileParser& par
   assert(loader_data != nullptr, "invariant");
 
   InstanceKlass* ik;
-  const bool use_class_space = !parser.is_interface() && !parser.is_abstract();
+  // Only Klasses that aren't interface or abstract classes are found in oop headers so they
+  // are the only ones that require compression and allocation in the limited class metaspace
+  // region.  But ... there are C2 dependencies that all klasses are allocated to the class metaspace.
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/instanceKlass.cpp:459]:
```suggestion
  // region.  But ... there are C2 dependencies that all klasses are allocated to the class metaspace, so
  // always use the class metaspace until JDK-NNNNNNN is resolved.
```

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
So this is just disabling [JDK-8338526](https://bugs.openjdk.org/browse/JDK-8338526) until other C2 changes are in place. Are they coming soon? If not shouldn't [JDK-8338526](https://bugs.openjdk.org/browse/JDK-8338526) be backed out?

**Review by tstuefe — APPROVED** [review]
Okay

**Comment by tstuefe** [comment]
> So this is just disabling [JDK-8338526](https://bugs.openjdk.org/browse/JDK-8338526) until other C2 changes are in place. Are they coming soon? If not shouldn't [JDK-8338526](https://bugs.openjdk.org/browse/JDK-8338526) be backed out?

[JDK-8338526](https://bugs.openjdk.org/browse/JDK-8338526) cannot be backed out verbatim, since it has some follow-up fixes. But I also don't think this is necessary at this point. The changes supporting this feature were harmless and few. And we still may want this feature, especially when it comes to auto-generated classes.

**Comment by TobiHartmann** [comment]
FTR, I had a look at [JDK-8343206](https://bugs.openjdk.org/browse/JDK-8343206) and I don't think it's an issue in current code.

**Comment by coleenp** [comment]
> So this is just disabling [JDK-8338526](https://bugs.openjdk.org/browse/JDK-8338526) until other C2 changes are in place. Are they coming soon? If not shouldn't [JDK-8338526](https://bugs.openjdk.org/browse/JDK-8338526) be backed out?

The changes for JDK-8338526 were minor refactoring and then there were nice follow up changes to test if klass pointers were encodable, as Thomas said above. The refactoring doesn't hurt anything and setting this way allows the opportunity for a comment explaining why, so we can repeat this change in the future.

> FTR, I had a look at [JDK-8343206](https://bugs.openjdk.org/browse/JDK-8343206) and I don't think it's an issue in current code.

Thank you Tobias for looking at this.  I'll answer in the bug report but I would feel better if the c2 code hit an assert or something if we tried to encode a Klass pointer that isn't in class-metaspace, so I think we should have my change to disable the allocation for now.

**Comment by coleenp** [comment]
I answered before reading JDK-8343206 where you do add a test for this.

File: src/hotspot/share/oops/instanceKlass.cpp [hunk 2]
```diff
@@ -454,7 +454,11 @@ InstanceKlass* InstanceKlass::allocate_instance_klass(const ClassFileParser& par
   assert(loader_data != nullptr, "invariant");
 
   InstanceKlass* ik;
-  const bool use_class_space = !parser.is_interface() && !parser.is_abstract();
+  // Only Klasses that aren't interface or abstract classes are found in oop headers so they
+  // are the only ones that require compression and allocation in the limited class metaspace
+  // region.  But ... there are C2 dependencies that all klasses are allocated to the class metaspace, so
+  // always use the class metaspace until JDK-NNNNNNN is resolved.
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/instanceKlass.cpp:460]: JDK-NNNNNN was supposed to be replaced by an actual bug number. :)

Comment (dholmes-ora) [reply src/hotspot/share/oops/instanceKlass.cpp:460]: I'm left wondering what this is supposed to do. We made changes, then we are disabling them until something is fixed but we are not saying what something that is. ???

Comment (coleenp) [reply src/hotspot/share/oops/instanceKlass.cpp:460]: Given that it's a subtask of the bug that we want to fix, it's pretty easy to see with git blame.  Like with all changes, we don't put bug numbers in.  But the bug will be fixed soon so I'm closing this as WNF.

**Comment by coleenp** [comment]
I don't see any evidence of c1 unsafely assuming that a Klass pointer can be encoded and decoded.  Since the parent CR for this bug is being fixed to add asserts for safety in C2 [JDK-8343206](https://bugs.openjdk.org/browse/JDK-8343206), I'm closing this as WNF.

**Pull request closed without merging**